pkcs11sec/pkcs11_security.go

```diff
@@ -0,0 +1,642 @@
+func (p *Pkcs11Security) VerifyCertificate(certpem []byte, name string) error {
+	return p.fsec.VerifyCertificate(certpem, name)
+}
+
+// PublicCertPem retrieves the public certificate for this instance
+func (p *Pkcs11Security) PublicCertPem() (*pem.Block, error) {
+
+	if p.pin == nil {
+		p.log.Debug("Attempting to login to token in PublicCertPem()")
+		if err := p.loginToToken(); err != nil {
+			return nil, errors.Wrap(err, "failed to login to token in PublicCertPem()")
+		}
+	}
+
+	pb := &pem.Block{
+		Type:  "CERTIFICATE",
+		Bytes: p.cert.Leaf.Raw,
+	}
+
+	return pb, nil
+}
+
+// PublicCertTXT retrieves pem data in textual form for the public certificate of the current identity
+func (p *Pkcs11Security) PublicCertTXT() ([]byte, error) {
+
+	pemCert, err := p.PublicCertPem()
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to run PublicCertPem")
+	}
+	var buf bytes.Buffer
+	if err := pem.Encode(&buf, pemCert); err != nil {
+		return nil, errors.Wrap(err, "failed to run pem.Encode")
+	}
+	return buf.Bytes(), nil
+}
+
+// Identity determines the choria certname
+func (p *Pkcs11Security) Identity() string {
+	if p.pin == nil {
+		p.log.Debug("Attempting to login to token in Identity()")
+		if err := p.loginToToken(); err != nil {
+			return "invalid"
+		}
+	}
+	return p.cert.Leaf.Subject.CommonName
+}
+
+// TLSConfig creates a TLS configuration for use by NATS, HTTPS etc
+func (p *Pkcs11Security) TLSConfig() (*tls.Config, error) {
+
+	caCert, err := ioutil.ReadFile(p.conf.CAFile)
+	if err != nil {
+		return nil, err
+	}
+
+	if p.pin == nil {
+		p.log.Debug("Attempting to login to token in TLSConfig()")
+		if err = p.loginToToken(); err != nil {
+			return nil, errors.Wrap(err, "failed to login to token in TLSConfig()")
+		}
+	}
+
+	caCertPool := x509.NewCertPool()
+	caCertPool.AppendCertsFromPEM(caCert)
+
+	tlsc := &tls.Config{
+		MinVersion:   tls.VersionTLS12,
+		Certificates: []tls.Certificate{*p.cert},
+		GetClientCertificate: func(*tls.CertificateRequestInfo) (*tls.Certificate, error) {
+			return p.cert, nil
+		},
+		ClientCAs: caCertPool,
+		RootCAs:   caCertPool,
+	}
+
+	if p.conf.DisableTLSVerify {
+		tlsc.InsecureSkipVerify = true
+	}
+
+	tlsc.BuildNameToCertificate()
+
+	return tlsc, nil
+}
+
+// SSLContext creates a SSL context loaded with our certs and ca
+func (p *Pkcs11Security) SSLContext() (*http.Transport, error) {
+	tlsConfig, err := p.TLSConfig()
+	if err != nil {
+		return nil, err
+	}
+
+	transport := &http.Transport{TLSClientConfig: tlsConfig}
+
+	return transport, nil
+}
+
+func (p *Pkcs11Security) HTTPClient(secure bool) (*http.Client, error) {
+	client := &http.Client{}
+
+	if secure {
+		tlsc, err := p.TLSConfig()
+		if err != nil {
+			return nil, fmt.Errorf("pkcs11: could not set up HTTP connection: %s", err)
+		}
+
+		client.Transport = &http.Transport{TLSClientConfig: tlsc}
+	}
+
+	return client, nil
+}
+
+// TODO: cache code from filesec should be pulled out/refactored
```

Comment:
the cache is only really used on the servers where presumably we dont use this security provider, so probably its ok without?

So these looks like they can proxy to the file one - but they're not exported there? Anyway as this is going to be basically client only useful we can ignore all the cache stuff

Reply:
We could potentially use pkcs11 on a server, which is why I left this in. You can imagine a server backed by an HSM. Would you mind if we kept the cache stuff in?

Reply:
I should add that there is an option helper for adding a pin that doesn't use stdin

Reply:
No I fond mine, then we need to remove the TODOs :)